tools/migration-unified-annotation-versioning/migration.py

```diff
@@ -0,0 +1,529 @@
+            self.save_bytes(collection, tracing_id, new_version, value_bytes)
+        return materialized_versions_unified
+
+    def list_versions(self, collection, key) -> List[int]:
+        reply = self.src_stub.ListVersions(proto.ListVersionsRequest(collection=collection, key=key))
+        assert_grpc_success(reply)
+        return reply.versions
+
+    def get_bytes(self, collection: str, key: str, version: int) -> bytes:
+        reply = self.src_stub.Get(proto.GetRequest(collection=collection, key=key, version=version))
+        assert_grpc_success(reply)
+        return reply.value
+
+    def save_bytes(self, collection: str, key: str, version: int, value: bytes) -> None:
+        if self.dst_stub is not None:
+            reply = self.dst_stub.Put(proto.PutRequest(collection=collection, key=key, version=version, value=value))
+            assert_grpc_success(reply)
+
+    def migrate_volume_buckets(self, tracing_id: str, layer_version_mapping: LayerVersionMapping):
+        self.migrate_all_versions_and_keys_with_prefix("volumeData", tracing_id, layer_version_mapping, transform_key=self.remove_morton_index)
+
+    def migrate_all_versions_and_keys_with_prefix(self, collection: str, tracing_id: str, layer_version_mapping: LayerVersionMapping, transform_key: Optional[Callable[[str], str]]):
+        list_keys_page_size = 5000
+        versions_page_size = 500
+        current_start_after_key = tracing_id + "."  # . is lexicographically before /
+        newest_tracing_version = max(layer_version_mapping[tracing_id].keys())
+        while True:
+            list_keys_reply = self.src_stub.ListKeys(proto.ListKeysRequest(collection=collection, limit=list_keys_page_size, startAfterKey=current_start_after_key))
+            assert_grpc_success(list_keys_reply)
+            if len(list_keys_reply.keys) == 0:
+                # We iterated towards the very end of the collection
+                return
+            for key in list_keys_reply.keys:
+                if key.startswith(tracing_id):
+                    for version_range_start, version_range_end in batch_range(newest_tracing_version, versions_page_size):
+                        get_versions_reply = self.src_stub.GetMultipleVersions(proto.GetMultipleVersionsRequest(collection=collection, key=key, oldestVersion=version_range_start, newestVersion=version_range_end))
+                        assert_grpc_success(get_versions_reply)
+                        new_key = key
+                        if transform_key is not None:
+                            new_key = transform_key(key)
+                        for version, value in zip(get_versions_reply.versions, get_versions_reply.values):
+                            if version not in layer_version_mapping[tracing_id]:
+                                continue
+                            new_version = layer_version_mapping[tracing_id][version]
+                            self.save_bytes(collection, new_key, new_version, value)
```

Comment:
Can't batch these writes? I guess that could be a big perf win.

Reply:
Probably true, but that would require changes to the FossilDB. Do you think I should go for that?

Reply:
If that is not too much effort in FossilDB, I think that could be useful. Not only for this migration but also for normal operation.

Reply:
I adapted the migration to make use of the new API introduced in https://github.com/scalableminds/fossildb/pull/47
Note that not all saves are covered by this, but the straightforward ones where multiple version/value pairs are processed in the same batch anyway. I think it might already give some speedup, but I did not re-test this at scale yet